added a prototype for par processing.

HryhoriiHevorkian opened the pull request:
In scope of this PR i would like to show the conception of PAR request processing. 
Please note, there is no check if the request is correct and no authorization process also.

If the conception is correct - nice (I will go on then), if there are some advices or propositions - feel free to comment. Thank you!

tnorimat commented:
@HryhoriiHevorkian Thank you. I would like to your codes and add comments afterwards.

tnorimat commented on services/src/main/java/org/keycloak/authentication/ParResponse.java:
```diff
@@ -0,0 +1,28 @@
+package org.keycloak.authentication;
```
It might be preferable to put PAR related classes on to its dedicated package. (ex. org.keycloak.protocol.par)

HryhoriiHevorkian replied:
Done, thank you.

tnorimat replied:
I've confirmed it. Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java:
```diff
@@ -33,13 +33,7 @@
-import org.keycloak.protocol.oidc.endpoints.TokenEndpoint;
-import org.keycloak.protocol.oidc.endpoints.TokenRevocationEndpoint;
-import org.keycloak.protocol.oidc.endpoints.UserInfoEndpoint;
+import org.keycloak.protocol.oidc.endpoints.*;
```
AFAIK, keycloak's coding style seems to be writing import directives one by one.

HryhoriiHevorkian replied:
Done.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java:
```diff
@@ -189,6 +183,16 @@ public Object token() {
+     * Pushed Authorization Request endpoint
+     */
+    @Path("par")
+    public Object par() {
```
As for the way of implementing Endpoint class, It might be perferable to follow `DeviceEndpoint` and `DeviceEndpointFactory`.

HryhoriiHevorkian replied:
I have changed this, thank you.

tnorimat replied:
I've confirmed the change of endpoint implementation style. Thank you.
IMO, `ParEndpoint` and `ParEndpointFactory` might be in `org.keycloak.protocol.par.endpoints` as the same manner in Device Grant's endpoint and its factory. WDYT?

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/ParEndpoint.java:
```diff
@@ -0,0 +1,76 @@
+    @POST
+    @NoCache
+    @Consumes({MediaType.APPLICATION_FORM_URLENCODED})
+    public Response handlePar(@FormParam(OBJECT_ID) String objectId,
```
Basically, this method has to receive and process parameters the same as ones in the current `AuthorizationEndpoint`. In addition, this method may receive some other parameters that are not defined by any standard documents like RFC 6749 and so on. [PAR's standard document](https://tools.ietf.org/html/draft-ietf-oauth-par-07#section-2.1) allows to support such extension (here, `object_id`, `object_type` and `username`). IMO, such extension have to be treated as Optional.

HryhoriiHevorkian replied:
I have implemented the map of parameters as the input data of this method. So, we can pass any key-value pairs.

tnorimat replied:
I've confirmed its change. Thank you.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/ParEndpoint.java:
```diff
@@ -0,0 +1,76 @@
+
+        LOG.info("Received PAR for object ID: " + objectId);
+
+        CodeToTokenStoreProvider codeStore = session.getProvider(CodeToTokenStoreProvider.class,
```
For the way of using `CodeToTokenStoreProvider`, it might be preferable to follow `OAuth2DeviceTokenStoreProvider` and `InfinispanOAuth2DeviceTokenStoreProvider` used for Device Grant.

HryhoriiHevorkian replied:
I have a question about this comment. The `OAuth2DeviceTokenStoreProvider ` uses much more functionality in comparison with the `CodeToTokenStoreProvider `. Basically, all we need is `put` and `get` methods, which are implemented in the `CodeToTokenStoreProvider `, we can store the map of parameters and then read this map only once (it is actually disposable). Do we really need the OAuth2DeviceTokenStoreProvider class?

tnorimat replied:
Yes, you are right. We need not to re-use `OAuth2DeviceTokenStoreProvider`.

IMO it might be better to implement PAR specific Store Provider which receive and return the POJO instance of PAR request and its key is `request_uri`.
```
    void put(String requestUri, int lifespanSeconds, ParRequestModel parRequest);
    ParRequestModel remove(String requestUri);
```
WDYT?

HryhoriiHevorkian replied:
Thank you for the advice! Yes, I have created a separate storage, factory and SPI for handling it. Those changes are already pushed to this branch, please, have a look.

tnorimat reviewed (commented):
@HryhoriiHevorkian I've added the review comments. Could you check them?

HryhoriiHevorkian commented:
@tnorimat  Hello. Could you please have a look at my last changes? I have implemented the validation and client authentication flow for PAR. Also I have added UI parameters - they are located in the "Advanced Settings" in the client configuration.

tnorimat commented:
@HryhoriiHevorkian Thank you. I would like to review your commits.

tnorimat reviewed (commented):
@HryhoriiHevorkian I've added some review comments. Could you check them?

mposolda reviewed (changes requested):
@HryhoriiHevorkian @tnorimat I've did just a brief scan of the PR. I've added some comments and I agree with most of comments from @tnorimat .

Besides that, it looks that this implementation is just a start, but it is not complete at all? Especially:
- ParEndpoint is not complete. It does not contain any implementations and checks, which are required by the specification it seems?
- There is not yet any support in the AuthorizationEndpoint (or AuthzEndpointRequestParser and similar classes) for the PAR requests parsing

HryhoriiHevorkian commented:
Hello, @tnorimat @mposolda
I have created a separate service class for PAR validation, please, have a look. Also, I have implemented all the recommendations provided by Takashi on the previous step. Thank you for your help.

mposolda commented:
@HryhoriiHevorkian Just a note, the implementation is still missing lots of stuff like:
- DescriptionConverter (Implementing configuration to/from OIDC config into ParConfig)
- AuthorizationEndpoint integration
- The ParValidationService : I think it is step in the good direction. I am not yet sure if have separate validation service or somehow re-use some validations from AuthorizationEndpoint. But maybe you can use ParValidationService for now and we can decide later if it is better to keep all the PAR related validations in separate class, like you are doing, or somehow try to re-use the validations...

tnorimat commented on services/src/main/java/org/keycloak/services/ParValidationService.java:
```diff
@@ -0,0 +1,159 @@
+
+        if (authorizationRequest.getInvalidRequestMessage() != null) {
+            event.error(Errors.INVALID_REQUEST);
+            return redirectErrorToClient(parsedResponseMode, Errors.INVALID_REQUEST, authorizationRequest.getInvalidRequestMessage());
```
I think that error codes might not be implemented properly. This method is for Authorization Endpoint. PAR does not relate to HTTP redirect.

tnorimat commented on services/src/main/java/org/keycloak/services/ParValidationService.java:
```diff
@@ -0,0 +1,159 @@
+        this.event = event;
+    }
+
+    public Response validateParRequest(HttpRequest request, ClientModel clientModel) {
```
According to https://datatracker.ietf.org/doc/html/draft-ietf-oauth-par#section-4 ,

>     3.  If the client has authentication credentials established with the
>       authorization server, reject the request if the authenticated
>       "client_id" does not match the "client_id" claim in the request
>       object.

It seems not to be implemented.

tnorimat reviewed (changes requested):
I've added some review comments. Could you check them?

mposolda commented on core/src/main/java/org/keycloak/representations/oidc/OIDCClientRepresentation.java:
```diff
@@ -487,4 +493,19 @@ public void setTlsClientAuthSubjectDn(String tls_client_auth_subject_dn) {
+        this.require_pushed_authorization_requests = require_pushed_authorization_requests;
+    }
+
+    public Integer getRequestUriLifespan() {
```
+1 . We discussed especially the "request_uri_lifespan" with @guymoyo in the design document and he mentioned that it is not mentioned in any specification and he will remove it from the design. I suggest to remove it from this PR as well and stick just to the options, which are mentioned in the specification https://datatracker.ietf.org/doc/html/draft-ietf-oauth-par#section-4

mposolda commented on server-spi-private/src/main/java/org/keycloak/models/PushedAuthzRequestStoreProviderFactory.java:
```diff
@@ -0,0 +1,23 @@
+
+import org.keycloak.provider.ProviderFactory;
+
+public interface PushedAuthzRequestStoreProviderFactory extends ProviderFactory<PushedAuthzRequestStoreProvider> {
```
Yes, that will be better. +1 to add PAR as a feature and have it "preview" by default similarly like CIBA

mposolda commented on server-spi/src/main/java/org/keycloak/models/ParConfig.java:
```diff
@@ -0,0 +1,11 @@
+
+    public static final String REQUIRE_PUSHED_AUTHORIZATION_REQUESTS = "require.pushed.authorization.requests";
+    public static final String REQUEST_URI_LIFESPAN = "request.uri.lifespan";
+
```
+1

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/ParEndpoint.java:
```diff
@@ -0,0 +1,190 @@
+/*
```
@tnorimat Yes, this package structure will be better. It will be nice if all "par" related classes are in separate package exactly as you proposed.

mposolda commented on services/src/main/java/org/keycloak/protocol/par/ParEndpointFactory.java:
```diff
@@ -0,0 +1,42 @@
+package org.keycloak.protocol.par;
```
+1

mposolda commented on services/src/main/java/org/keycloak/protocol/par/ParResponse.java:
```diff
@@ -0,0 +1,33 @@
+package org.keycloak.protocol.par;
```
+1

HryhoriiHevorkian commented:
Hello, @tnorimat @mposolda 
I have a question regarding the request URI building. The documentation https://datatracker.ietf.org/doc/html/draft-ietf-oauth-par#section-2.2 says, that:

> The authorization server MAY construct the "request_uri" value using the form "urn:ietf:params:oauth:request_uri:<reference-value>" with "<reference-value>" as the random part of the URI that references the respective authorization request data.
> 
>    The "request_uri" value MUST be bound to the client that posted the
>    authorization request.

I understood this like the following: I have to calculate some kind of hash (?) from the form parameter "client_id" field, and then use this hash as the <reference-value> (maybe, encoded with Base64Url). Am I right?

Thank you, guys, for your help.

tnorimat commented:
Hello @HryhoriiHevorkian ,

IMO, the aim of this binding request_uri with client is to prevent authz server from retrieving other client's request on authorization endpoint.

The way you've mentioned is one way of realizing this binding. Considering that everyone can know other clients' cliend_id so that they can calculate other client's hash value, we need to include this hash value onto as input to derive the infinispan cache entry's key for holding the client's PAR request. If not, someone can replace client_id's hash value part of other client's request_uri with their client_id's hash value, which may lead to session hijacking or session fixation.

Alternative way is to store client_id showing the client that keycloak authenticated on PAR endpoint onto the infinispan cache entry for holding this client's PAR request. When keycloak receives the request with request_uri and client_id parameter on Authz Endpoint, keycloak retrieve the PAR request based on request_uri, get client_id enclosed in it, and compare it with the client_id parameter of authorization request on Authz Endpoint.

@mposolda  WDYT? or have other ways?

HryhoriiHevorkian commented:
Hello, @tnorimat 
Another question is interesting for me now. The second part of PAR processing: so, imagine, that we have already obtained the request_uri after the PAR request. Say, its value is `urn:ietf:params:oauth:request_uri:vmeCLb2L4LVpDN4Tlezokv5nFqbrq593ryWIvVpvGNk`. I use this value as the query parameter in usual authorization GET request. Then, inside the Keycloak's AuthorizationEndpointRequestParserProcessor class I see the following code:

>                 String requestUri = RedirectUtils.verifyRedirectUri(session, client.getRootUrl(), requestUriParam, new HashSet<>(requestUris), false);
>                 if (requestUri == null) {
>                     throw new RuntimeException("Specified 'request_uri' not allowed for this client.");
>                 }
> 
>                 try (InputStream is = session.getProvider(HttpClientProvider.class).get(requestUri)) {
>                     String retrievedRequest = StreamUtil.readString(is);
>                     new AuthzEndpointRequestObjectParser(session, retrievedRequest, client).parseRequest(request);
>                 }
> 

Request URI is validated, it is not null, everything is fine and we try to read an InputStream from this URI. But the URI (which is actually, URN) can not load this InputStream. IMO, we need to implement some kind of `if then` block before this InputStream reading, and in case of PAR is active, we have to load the PAR data from the StoreProvider. Am I right? Thank you!

tnorimat commented:
@HryhoriiHevorkian Yes, as you mentioned, we need to distinguish whether `request_uri`'s value is originated from PAR or Request Object.

HryhoriiHevorkian commented:
Hello, @tnorimat 
So, I have implemented the request type distinguishing as we discussed - you may see the `private static AuthorizationEndpointRequest processPar(KeycloakSession session, String requestUri)` method inside the `AuthorizationEndpointRequestParserProcessor` class. The next question which is not described in the specification: what should we do with the map of PAR request parameters after obtaining them? They are available as the map of key-value string parameters, but they can not be useful inside the Authorization flow. IMO we have to set them as custom claims inside the token, but in such case we have to modify the token endpoint REST controller (and the token flow, also) in a way that it is ready to receive the `request_uri` parameter. Is that correct? Or there should be other variant of PAR parameters processing? Thank you!

tnorimat commented:
Hello @HryhoriiHevorkian,

I think that authorization request parameters as PAR are used in an authorization flow afterwards the same as request object and query parameters of http redirect.

All PAR, request object and query parameters of HTTP redirect convey the same authorization request parameters.
For example, client_id, redirect_uri, code_challenge, code_challenge_method=S256, scope, and so on.

Then, there are two types of authorization request parameters, ones defined by standard specification like RFC 6749, and the others by non-standard specification like use-case specific.

IMO, it might be better to parse authorization request parameters by PAR the same as in request object and query parameters of http redirect, and eventually put the result onto `AuthorizationEndpointRequest` and `AuthenticatedClientSessionModel`'s notes that can be looked by an authorization flow context afterwards.

How to handle authorization parameters by non-standard specification is not defined by any standard document, it depends on case by case.

HryhoriiHevorkian commented:
Hello, @tnorimat 
Please, check the last commit - is it the just idea you told me about? Thank you.

HryhoriiHevorkian commented:
And also the next question: current PAR happy-path flow results in the 200 HTTP response and HTML page as the response for authorization endpoint (addressing GET request to the `http://localhost:8081/auth/realms/master/protocol/openid-connect/auth `). It is not correct, we just deal with pure backend IMO, how should we prepare the response for this happy-path flow? \

Please have a look at the screenshot and the reply in my Postman:

<img width="1269" alt="Screen Shot 2021-05-25 at 10 42 01 AM" src="https://user-images.githubusercontent.com/64079106/119459343-42bd6180-bd46-11eb-9e8c-127f5f41a292.png">

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java:
```diff
@@ -448,6 +449,10 @@ private void updateAuthenticationSession() {
         authenticationSession.setClientNote(OIDCLoginProtocol.REDIRECT_URI_PARAM, request.getRedirectUriParam());
         authenticationSession.setClientNote(OIDCLoginProtocol.ISSUER, Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
 
+        for (Map.Entry<String, String> entry : request.getAdditionalReqParams().entrySet()) {
```
This method has already stored additional parameters to `AuthenticationSessionModel`.
I'm afraid this code does the same thing.
```
        if (request.getAdditionalReqParams() != null) {
            for (String paramName : request.getAdditionalReqParams().keySet()) {
                authenticationSession.setClientNote(LOGIN_SESSION_NOTE_ADDITIONAL_REQ_PARAMS_PREFIX + paramName, request.getAdditionalReqParams().get(paramName));
            }
        }
```

tnorimat replied:
There might be a chance that `request.getAdditionalReqParams()` is null.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthorizationEndpointRequestParserProcessor.java:
```diff
@@ -109,4 +128,33 @@ public static String getClientId(EventBuilder event, KeycloakSession session, Mu
+
+        if (System.currentTimeMillis() - created < (expiresIn * MILLIS_IN_SECOND)) {
+            // happy path - process PAR.
+            for (Map.Entry<String, String> entry : retrievedRequest.entrySet()) {
```
There might be a chance that `retrievedRequest` is null.

tnorimat reviewed (commented):
@HryhoriiHevorkian I've reviewed your last commit and add some review comments. Could you check them?

HryhoriiHevorkian commented:
Hello, @tnorimat 
I have added hashing for client ID parameter - could you please have a look at the changes? And, also, please see my question above regarding the response for authorization with PAR in case of happy-path.
Thank you and have a nice day.

tnorimat commented:
@HryhoriiHevorkian Hello,
As for https://github.com/tnorimat/keycloak/pull/20#issuecomment-847633360 , does you means that you want to not follow an authorization code grant?

AFAIK, the aim of PAR is for registering authorization request parameters in advance so that we can avoid putting authorization request parameters as query parameters in an authorization code grant.

HryhoriiHevorkian commented:
@tnorimat 
@mposolda 

Hello, could you please have a look at the documentation? I have attached the txt file, which is actually adoc format.

[par.txt](https://github.com/tnorimat/keycloak/files/6570844/par.txt)

tnorimat commented:
Hello @HryhoriiHevorkian ,
I've read the text. It make sense. Thank you for your clarification.

As for your problem https://github.com/tnorimat/keycloak/pull/20#issuecomment-847633360 ,
I'm now investigating it. I've found some points we need to resolve but I'll tell you the complete result of this investigation afterwards.

HryhoriiHevorkian commented:
Hello, @tnorimat 
Thanks for reading the documentation. As for the problem I described: I have a definite result already and it is successful. So, here are the steps you should implement to get the token with my new PAR functionality:

- Launch keycloak with PAR feature activated -Dkeycloak.profile.feature.par=enabled
- Login as an admin (default credentials are "admin"/"admin")
- Create new client - “par-client”
- Set its access type to “confidential”
- Set the valid Redirect URI, e.g. “www.google.com”
- Go to “Fine Grain OpenID Connect Configuration” and set “Valid Request URIs” to “*”
- Go to “ Advanced Settings “ and switch on the “Pushed Authorization Request Enabled “
- Go to “Realm settings” - ‘Tokens” and set the “Lifetime of the Request URI for Pushed Authorization Request ” to 10 minutes
- Create a new user with password, e.g. “anton.brueckner” and “12345”
- Run PAR request with parameters (POST http://localhost:8081/auth/realms/master/par, parameters are: client_id=par-client, client_secret=xxxxxxxx, redirect_uri=www.google.com, response_type=code and you may adding custom parameters you wish) and copy the request_uri from the response
- Run the authorization request from the browser with request URI in the path (GET http://localhost:8081/auth/realms/master/protocol/openid-connect/auth?client_id=par-client&request_uri=urn:ietf:params:oauth:request_uri:U2SgHngW9_vcB3GWPqtuIiUyWQur_LrfuAgYbjVOb1O4VKqBhy4bajymtBOiqiCBsIaQyve4wODNiGtL8ipT8A&response_type=code)
- The browser will display you the login window, at this step open the browser console to see the network activity
- In this window provide your user’s credentials (anton.brueckner and 12345)
- Keycloak would display “We are sorry…“, but at the console you can find the failed request with “www.google.com” redirect URI and this GET request contains the code we need in the pass: “…….code=3d7103b7-f038-4f6b-8976-9cab5e8e8f81.5cf8ae16-dd2c-4b5e-8b10-095764c42980.7c01ff3e-38e6-48cd-8ce8-7621727c8577”
- Copy this code
- Final step - run the token request (POST http://localhost:8081/auth/realms/master/protocol/openid-connect/token with parameters: client_id=par-client, client_secret=xxxxxxxx, grant_type=authorization_code, code=3d7103b7-f038-4f6b-8976-9cab5e8e8f81.5cf8ae16-dd2c-4b5e-8b10-095764c42980.7c01ff3e-38e6-48cd-8ce8-7621727c8577, redirect_uri=www.google.com)
- Bingo - we have finally obtained the token as a response! 
 
The only failing step here is wrong redirect URI concatenating, just because of this we have to open the browser console and see the request internals to obtain the code. Everything else seems to work as expected. Maybe, you can give some advices about the redirect URI building? Many thanks, have a nice day.

tnorimat commented:
@HryhoriiHevorkian I could resolve your redirect_uri problem. I will tell you it after my investigation is completed.

Also, to accelerate this work on upstreaming this PAR PR, I could take over writing Arquillian Integration Tests that is needed when you send this PR to keycloak upstream. You can concentrate on the PAR's feature codes of this PR. WDYT?

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/par/endpoints/ParEndpoint.java:
```diff
@@ -0,0 +1,201 @@
+
+        int expiresIn = realm.getAttribute("requestUriLifespan", 60);
+
+        request.getFormParameters().forEach((k, v) -> params.put(k, String.valueOf(v)));
```
`request.getFormParameters()` returns percent encoded redirect_uri.
`request.getDecodedFormParameters()` may resolve the redirect_uri related problem you've encountered.

HryhoriiHevorkian replied:
@tnorimat this is not the issue, let me please describe it more precisely. First, I have provided redirect URI in my client configuration - see the picture 1:

<img width="569" alt="Screen Shot 2021-06-02 at 3 28 33 PM" src="https://user-images.githubusercontent.com/64079106/120480042-4d679e80-c3b7-11eb-94f8-f71df039c96a.png">

After launching the whole flow and providing the user's credentials, I can see the following in the browser console (picture 2):

![wrong1](https://user-images.githubusercontent.com/64079106/120480202-75ef9880-c3b7-11eb-855a-778aa9b1e913.png)

That is why this request fails with 404 and the error “We are sorry…“ is displayed. Please note, that the link is built in a **proper** way (without the "http://localhost:8081/auth/" prefix) in case of providing the redirect URI parameter "localhost:8088", even if there is no application on my machine at the 8088 port. Is there a way to switch off the "http://localhost:8081/auth/" prefix during the link building? Thanks, have a nice day.

HryhoriiHevorkian commented:
> @HryhoriiHevorkian I could resolve your redirect_uri problem. I will tell you it after my investigation is completed.
> 
> Also, to accelerate this work on upstreaming this PAR PR, I could take over writing Arquillian Integration Tests that is needed when you send this PR to keycloak upstream. You can concentrate on the PAR's feature codes of this PR. WDYT?

Yes, you may start writing the tests, thank you.

tnorimat commented:
@HryhoriiHevorkian 
It seems that you've set  `www.google.com` to `Varid Redirect URIs` on keycloak's Admin Console.
Also, you've put `redirect_uri` as follows when you send PAR to keycloak.
```
redirect_uri=www.google.com
```

How about the following?
1. On keycloak's admin console, set  `https://www.google.com` to `Varid Redirect URIs`.
1. On sending PAR to keycloak, put `redirect_uri` as follows when you send PAR to keycloak.
```
redirect_uri=https%3A%2F%2Fwww.google.com
```

tnorimat commented:
Hello @HryhoriiHevorkian ,

I've added `ParTest` of Arquillian integration tests.
https://github.com/tnorimat/keycloak/tree/added_par_endpoint

I've run this test and found some problems. I've also fixed them. Here I would like tell you how to fix them.
https://github.com/tnorimat/keycloak/compare/700754e75b5de81b0af884a0df9f59fe46ee93ee...72ef5ac3025904ace2fa57d149fa24c733c810f7

- Revise `OIDCWellKnownProviderTest` because PAR's server metadata are introduced.
- Change how to parse PAR on PAR Endpoint : use `getDecodedFormParameters()` and strip `[` and `]` here.
- Change how to parse PAR on Authz Endpoint : create `AuthzEndpointParParser` to follow the style of parsing other type authz request (query string for `AuthzEndpointQueryStringParser` and request object for `AuthzEndpointRequestObjectParser`).
- Add header comments for copyright.

I've checked the following test by `ParTest`. One test cannot work well. I've now being investigating it. 

- [x] success with one client conducting one authz request with PAR
- [x] success with the same client conducting multiple authz requests with PAR simultaneously 
- [x] success with the same client conducting multiple authz requests + PAR simultaneously
- [x] fail when not issued PAR is used
- [x] fail when the same PAR is used twice  
- [x] fail when PAR is used by other client
- [x] fail when PAR is used by not allowed client
- [ ] fail when expired PAR is used

HryhoriiHevorkian commented:
Hello, @tnorimat 
I have updated my branch with your changes (including the test), and it seems to work OK using the "https://www.google.com" as a redirect URI. I'm redirected to the Google search page and in the address bar I see "&code=xxxxxxxx" (just what we expected). This code can be exchanged for token at the next step.

As for the tests - is there a way to launch them in debug mode somehow except this command?

`mvn -f testsuite/integration-arquillian/pom.xml clean install -Dtest=ParTest -Dmaven.surefire.debug=true
`

tnorimat commented:
Hello @HryhoriiHevorkian ,

Unfortunately, I don't know other ways of attaching keycloak during testing.
Also, I've confirmed that `testFailureParExpired()` could pass.

If you've finished coding features for PAR, it might be good to sent this PR to keycloak upstream.

HryhoriiHevorkian commented:
Hello, @tnorimat 

So, I have created an issue: https://issues.redhat.com/browse/KEYCLOAK-18353
Currently working on this branch and squashing the commits.

tnorimat commented:
Hello @HryhoriiHevorkian ,

Technically, you can send this PR to keycloak upstream. IMO, it is good to send this PR by youself because you've written almost all of codes of this PR.

However, if your company policy or other rules does not allow you to send this PR to keycloak upstream, I will do it and respond with keycloak's dev team's review on behalf of you. In this case, I will add you to `Author: ` trailer and me to `Co-authored-by: ` trailer.

```
KEYCLOAK-18353 Implement Pushed Authorization Request inside the Keycloak
Author : Hryhorii Hevorkian <hhe@adorsys.com.ua>

Co-authored-by: Takashi Norimatsu <takashi.norimatsu.ws@hitachi.com>
```

Which do you like better, sending this PR to keycloak upstream by yourself, or by me on behalf of you.

HryhoriiHevorkian commented:
Hello, @tnorimat
I think, it will be better that you push this PR to upstream and mention me as an author. Thank you in advance and also thank you for helping with the code of Keycloak core.

tnorimat commented:
Hello @HryhoriiHevorkian ,

I see, I will send PR to keycloak upstream on behalf of you. Thank you.

tnorimat commented:
Hello @HryhoriiHevorkian ,

I've been preparing upstreaming PR.
https://github.com/tnorimat/keycloak/tree/KEYCLOAK-18353

I have one point I would like to mention about PAR endpoint URL.

According to the design document and this PR's implementation, PAR endpoint URL (on master realm) is as follows.

```
"pushed_authorization_request_endpoint":
 "http://localhost:8081/auth/realms/master/par"
```

However, in current keycloak, other endpoints' URLs for OIDC are the following style.
```
"authorization_endpoint":
 "http://localhost:8081/auth/realms/master/protocol/openid-connect/auth",
"token_endpoint":
 "http://localhost:8081/auth/realms/master/protocol/openid-connect/token"
"revocation_endpoint":
 "http://localhost:8081/auth/realms/master/protocol/openid-connect/revoke"
"introspection_endpoint":
 "http://localhost:8081/auth/realms/master/protocol/openid-connect/token/introspect"
"jwks_uri":
 "http://localhost:8081/auth/realms/master/protocol/openid-connect/certs"
 "device_authorization_endpoint":
 "http://localhost:8081/auth/realms/master/protocol/openid-connect/device"
```

For SAML
```
 "http://localhost:8081/auth/realms/master/protocol/saml"
```

To follow these style, I've changed PAR endpoint URL as follows.

```
"pushed_authorization_request_endpoint":
 "http://localhost:8081/auth/realms/master/protocol/openid-connect/par"
```

If you have some problem on this change, please tell me.

tnorimat commented:
@HryhoriiHevorkian  I've sent PR to keycloak upstream https://github.com/keycloak/keycloak/pull/8144 so that I will close it.
